src/tqec/compile/observables/abstract_observable.py

```diff
@@ -0,0 +1,215 @@
+"""Describe the location of the observable measurements in the block graph."""
+
+from dataclasses import dataclass
+
+from tqec.compile.specs.enums import JunctionArms
+from tqec.computation.correlation import CorrelationSurface
+from tqec.computation.cube import Cube, ZXCube
+from tqec.computation.pipe import Pipe
+from tqec.position import Direction3D, Position3D
+from tqec.computation.block_graph import BlockGraph
+from tqec.computation.zx_graph import ZXEdge, ZXKind
+
+
+@dataclass(frozen=True)
+class AbstractObservable:
+    """An abstract description of a logical observable in the
+    :py:class:`~tqec.computation.BlockGraph`.
+
+    A logical observable corresponds to all the
+    `OBSERVABLE_INCLUDE <https://github.com/quantumlib/Stim/blob/main/doc/gates.md#OBSERVABLE_INCLUDE>`_
+    instructions with the same observable index in the ``stim`` circuit
+    and is composed of a set of measurements. Abstract observable
+    specifies where are the measurements located in the block graph.
+
+    Attributes:
+        top_readout_cubes: A set of cubes of which a straight line of data
+            qubit readouts on the top face should be included in the observable.
+        top_readout_pipes: A set of pipes of which a single data qubit readout
+            on the top face should be included in the observable.
```

Comment:
Maybe add explicitly that only one data-qubit from the pipe need to be included (and it is the central one) because data-qubits on the borders of the pipe are shared with neighbouring cubes, and the data-qubits are already included by the cube-processing functions?